bench/randindex_bench.d

```diff
@@ -0,0 +1,379 @@
+randIndexV1: 1.29997 * 10 ^^ 8 calls/s
+randIndexV2: 1.8269 * 10 ^^ 8 calls/s
+new mir.random.randIndex (potential inlining shenanigans): 2.30216 * 10 ^^ 8 calls/s
+
+Benchmarks for generating uint from XorshiftStarEngine!(ulong, 1024u, 31u, 11u, 30u, 11400714819323198483LU, ulong):
+randIndexV1: 2.28637 * 10 ^^ 8 calls/s
+randIndexV2: 3.16081 * 10 ^^ 8 calls/s
+new mir.random.randIndex (potential inlining shenanigans): 3.70199 * 10 ^^ 8 calls/s
+
+Benchmarks for generating uint from Xoroshiro128Plus:
+randIndexV1: 2.56082 * 10 ^^ 8 calls/s
+randIndexV2: 4.37158 * 10 ^^ 8 calls/s
+new mir.random.randIndex (potential inlining shenanigans): 6.9869 * 10 ^^ 8 calls/s
+
+Benchmarks for generating uint from PermutedCongruentialEngine!(xsh_rr, cast(stream_t)2, true):
+randIndexV1: 1.50404 * 10 ^^ 8 calls/s
+randIndexV2: 2.32221 * 10 ^^ 8 calls/s
+new mir.random.randIndex (potential inlining shenanigans): 3.86847 * 10 ^^ 8 calls/s
+
+Benchmarks for generating uint from PermutedCongruentialEngine!(rxs_m_xs_forward, cast(stream_t)2, true):
+randIndexV1: 1.47411 * 10 ^^ 8 calls/s
+randIndexV2: 2.49377 * 10 ^^ 8 calls/s
+new mir.random.randIndex (potential inlining shenanigans): 4.24178 * 10 ^^ 8 calls/s
+*/
+
+T randIndexV1(T, G)(ref G gen, T m)
+    if(isSaturatedRandomEngine!G && isUnsigned!T)
+{
+    pragma(inline, false);//Try to prevent LDC from doing anything clever with the modulus.
+
+    assert(m, "m must be positive");
+    T ret = void;
+    T val = void;
+    do
+    {
+        val = gen.rand!T;
+        ret = val % m;
+    }
+    while (val - ret > -m);
+    return ret;
+}
+
+T randIndexV2(T, G)(ref G gen, T m)
+    if(isSaturatedRandomEngine!G && isUnsigned!T)
+{
+    pragma(inline, false);//Try to prevent LDC from doing anything clever with the modulus.
+    static if (EngineReturnType!G.sizeof >= T.sizeof * 2)
+        alias MaybeR = EngineReturnType!G;
+    else static if (uint.sizeof >= T.sizeof * 2)
+        alias MaybeR = uint;
+    else static if (ulong.sizeof >= T.sizeof * 2)
+        alias MaybeR = ulong;
+    else static if (is(ucent) && __traits(compiles, {static assert(ucent.sizeof >= T.sizeof * 2);}))
+        mixin ("alias MaybeR = ucent;");
+    else
+        alias MaybeR = void;
+
+    version (LDC) if (!__ctfe)
+    {
+        static if (!is(MaybeR == void))
+        {
+            alias R = MaybeR;
+            static assert(R.sizeof >= T.sizeof * 2);
+            import mir.ndslice.internal: _expect;
+            //Use Daniel Lemire's fast alternative to modulo reduction:
+            //https://lemire.me/blog/2016/06/30/fast-random-shuffling/
+            R randombits = cast(R) gen.rand!T;
+            R multiresult = randombits * m;
+            T leftover = cast(T) multiresult;
+            if (_expect(leftover < m, false))
```

Comment:
OK, then `if` should have most probable condition:
```
...
if (leftover >= m)
    return cast(T) (multiresult >>> finalshift);
...
while(...)
```

Reply:
Is it really a problem to use _expect in runtime code only? We have a unittest that verifies that CTFE and runtime evaluation produce the same result.